Raptor integration - feature branch (#3121)
ezsystems/developer-documentation: docs/recommendations/raptor_integration/tracking_functions.md

---
description: Integrate the tracking script to collect user interactions.
month_change: true
---

# Raptor tracking functions

> Raptor [tracking functions](https://content.raptorservices.com/help-center/introduction-to-tracking-documentation) introduce visit tracking functionality for collecting user interactions with products and content.

Reviewer: ```suggestion
[Raptor connector](raptor_connector.md) introduces [visit tracking functionality](https://content.raptorservices.com/help-center/introduction-to-tracking-documentation) for collecting user interactions with products and content.
```

Raptor integration introduces two Twig functions:

- [`ibexa_tracking_script()`](../../templating/twig_function_reference/recommendations_twig_functions.md#ibexa_tracking_script-function) - allows you to embed main tracking script into the website.
- [`ibexa_tracking_track_event()`](../../templating/twig_function_reference/recommendations_twig_functions.md#ibexa_tracking_track_event-function) - is responsible for sending event data to the service, enabling tracking of user interactions and behaviors.

## Embed tracking script

To enable tracking, tracking script must be embedded into the website’s layout.
To embed tracking script, add the twig function `ibexa_tracking_script()` into the <head> section of your base layout template, for example, `pagelayout.html.twig`:

``` html+twig
[[= include_file('code_samples/recommendations/templates/themes/standard/pagelayout.html.twig') =]]
```

## Tracking modes

Tracking user interactions can be implemented on the client-side or the server-side.
Each approach differs in where events are captured and how they are sent to the tracking backend.

> The tracking function outputs different content depending on the mode:

Reviewer: ```suggestion
The [tracking Twig function](#embed-tracking-script) outputs different content depending on the mode:
```

---
description: Integrate the tracking script to collect user interactions.
month_change: true
---

# Raptor tracking functions

[Raptor connector](raptor_connector.md) introduces [visit tracking functionality](https://content.raptorservices.com/help-center/introduction-to-tracking-documentation) for collecting user interactions with products and content.
The implementation includes product visit tracking with mapping to tracking parameters, and Twig functions for straightforward integration.

Raptor integration introduces two Twig functions:

- [`ibexa_tracking_script()`](../../templating/twig_function_reference/recommendations_twig_functions.md#ibexa_tracking_script-function) - allows you to embed main tracking script into the website.
- [`ibexa_tracking_track_event()`](../../templating/twig_function_reference/recommendations_twig_functions.md#ibexa_tracking_track_event-function) - is responsible for sending event data to the service, enabling tracking of user interactions and behaviors.

## Embed tracking script

To enable tracking, tracking script must be embedded into the website’s layout.

> To embed tracking script, add the twig function `ibexa_tracking_script()` into the <head> section of your base layout template, for example, `@ibexadesign/pagelayout.html.twig`:

Reviewer 2: ```suggestion
You must embed the tracking script into the website’s layout to enable tracking.
To do it, add the `ibexa_tracking_script()` twig function into the <head> section of your base layout template, for example, `@ibexadesign/pagelayout.html.twig`:
```

``` html+twig
[[= include_file('code_samples/recommendations/templates/themes/standard/pagelayout.html.twig') =]]
```

## Tracking modes

Tracking user interactions can be implemented on the client-side or the server-side.
Each approach differs in where events are captured and how they are sent to the tracking backend.

The [tracking Twig function](#embed-tracking-script) outputs different content depending on the mode:

``` yaml
# Server-side tracking
connector_raptor:
    tracking_type: 'server'  # Returns HTML comments

# Client-side tracking
connector_raptor:
    tracking_type: 'client'  # Returns <script> tags
```

> - **server** - returns HTML comments, placeholders that do not perform any action. Instead, the tracking is done server-side.

Reviewer 3: This is not precise. Server-side doesn't return any content for production environment. For development environments (or more precisely for these environments where `kernel.debug` Symfony container parameter is set to `true`) it outputs HTML comments.

Reviewer: @julitafalcondusza I would propose to answer @alongosz concern with this:
````suggestion
    tracking_type: 'server'  # Returns nothing (prod) or HTML comments (dev)

# Client-side tracking
connector_raptor:
    tracking_type: 'client'  # Returns <script> tags
```

- **server** - the tracking is done server-side. It doesn't return anything in `prod` environment, and returns HTML comments in `dev` environment or any environment where `kernel.debug` is `true`.
````
I also passed the main information to pole position: "done server-side"

|Template|Description|Example project path|
|--------|-----------|--------------------|
|`@ibexadesign/ibexa/tracking/script.html.twig`|Responsible for creating the `window.raptor` object and handling consent. Loads tracking only if consent is given and listens for the `enableTracking` event.|`templates/themes/standard/ibexa/tracking/script.html.twig`|
|`@ibexadesign/ibexa/tracking/script.js.twig`|Handles the loading of the tracking JavaScript.|`templates/themes/standard/ibexa/tracking/script.js.twig`  |

Available variables are:

- **customer_id** - type: string, Raptor account ID used for tracking
- **script_url** - type: string, URL of the tracking script, by default `//deliver.raptorstatic.com/script/raptor-3.0.min.js`, configurable through `ibexa.connector.raptor.tracking_script.url` Symfony Dependency Injection container parameter (not SiteAccess-aware)
- **has_consented** - type: boolean, indicates whether the user has given consent, default value: `false` (unless explicitly passed as function argument)
- **debug** - type: boolean, `kernel.debug` Symfony dependency injection container parameter, typically `true` in development environments and `false` in production

> The default template defines a Twig block that includes `script.js.twig`.

Reviewer 4: ```suggestion
The default template defines a `ibexa_tracking_script` Twig block that includes the `script.js.twig` template.
```

When extending the template, this block can be overridden to customize the script’s behavior.

You can override the default templates, either individually or both at the same time.

> ## Extension

Reviewer 4: ```suggestion
## Extending default templates
```

It's possible to extend `script.html.twig` by combining the [[= product_name_base =]] Design Engine with standard Symfony template reference in `templates/themes/standard/ibexa/tracking/script.html.twig`:

``` html+twig
[[= include_file('code_samples/recommendations/templates/themes/standard/ibexa/tracking/script.html.twig') =]]
```

> In most cases, the preferred approach is to do the opposite:

Reviewer 4: I thought we talked about not recommending any approach, just documenting what's possible? Just asking to be sure

Reviewer 5: I suggest changing this to: "As an alternative approach, you can override the template by providing a custom template path:"
